### PR TITLE
Heap snapshots

### PR DESCRIPTION
**Pull request by ewindisch**
Adds heap snapshotting. Uploads to the signed URL are now
zip files (this requires a synchronized change to manage
these files on the signing API!)

Signed-off-by: Erica Windisch <erica@iopipe.com>

### CODE REVIEW

**Comment by ewindisch** [comment]
Please review. Needs changes in upstream API before release.

**Comment by ewindisch** [comment]
Please review. Needs changes in upstream API before release.

@pselle @coreylight

File: src/enabled.js [hunk 1]
```diff
@@ -1,7 +1,17 @@
...
     enabledStatus ||
     (process.env.IOPIPE_ENABLE_PROFILER === 'true' ||
       process.env.IOPIPE_ENABLE_PROFILER === true)
   );
 }
+
+function getHeapSnapshotEnabledStatus(enabledStatus) {
+  return (
+    enabledStatus ||
+    (process.env.IOPIPE_ENABLE_HEAPSNAPSHOT === 'true' ||
```
Comment (coreylight) [inline src/enabled.js]: Consider namespacing environment variables per the plugin, i.e. `process.env.IOPIPE_PROFILER_ENABLE_HEAP`

File: src/index.js [hunk 2]
```diff
@@ -56,7 +63,9 @@ class ProfilerPlugin {
     this.log(`Requesting signed url from ${hostname}`);
     const signingRes = await request(
       JSON.stringify({
-        arn: context.invokedFunctionArn,
+        arn:
```
Comment (coreylight) [inline src/index.js]: Remove code (was used for debug)

File: src/enabled.js [hunk 3]
```diff
@@ -1,7 +1,17 @@
...
+      process.env.IOPIPE_ENABLE_HEAPSNAPSHOT === true)
+  );
+}
+
+module.exports = { getProfilerEnabledStatus, getHeapSnapshotEnabledStatus };
```
Comment (coreylight) [inline src/enabled.js]: Use es6 exports here ```export {getProfilerEnabledStatus, getHeapSnapshotEnabledStatus}```

File: src/index.js [hunk 4]
```diff
@@ -74,18 +83,40 @@ class ProfilerPlugin {
...
+          const profile = v8profiler.stopProfiling();
+          archive.append(profile.export(), { name: 'profile.cpuprofile' });
+        }
+        if (enabled.getHeapSnapshotEnabledStatus(this.heapsnapshotEnabled)) {
+          const heap = v8profiler.takeSnapshot();
+          archive.append(heap.export(), { name: 'profile.heapsnapshot' });
+        }
+        archive.finalize();
+      } catch (e) {
+        console.log('@iopipe/profiler::Error in upload:', e);
```
Comment (ewindisch) [inline src/index.js]: use this.log

File: src/index.js [hunk 5]
```diff
@@ -37,15 +44,15 @@ class ProfilerPlugin {
       name: pkg.name,
       version: pkg.version,
       homepage: pkg.homepage,
-      enabled: this.enabled
+      enabled: this.profilerEnabled || this.heapsnapshotEnabled
```
Comment (coreylight) [inline src/index.js]: use camelCase for `this.heapSnapshotEnabled`

File: src/index.js [hunk 6]
```diff
@@ -74,18 +83,40 @@ class ProfilerPlugin {
...
+    return new Promise(async resolve => {
+      try {
+        const signedRequestURL = await this.getSignedUrl();
+        const archive = archiver.default('zip');
+        const archiveBuffer = [];
+
+        archive.on('data', chunk => {
+          archiveBuffer.push(chunk);
+        });
+        archive.on('finish', async () => {
```
Comment (ewindisch) [inline src/index.js]: Note why we cannot use streams here (cannot do PUT into S3 signed urls with chucked encoding as a Stream would enforce)

File: src/index.js [hunk 7]
```diff
@@ -2,15 +2,17 @@ import v8profiler from 'v8-profiler-lambda';
 import * as urlLib from 'url';
 import get from 'lodash.get';
 import request from './request';
-import getEnabledStatus from './enabled';
+import * as enabled from './enabled';
 import getSignerHostname from './signer';
+import * as archiver from 'archiver';
```
Comment (coreylight) [inline src/index.js]: Move external dependencies above internal files. archiver before `./request`

File: __mocks__/v8-profiler-lambda.js [hunk 8]
```diff
@@ -1,8 +1,18 @@
+const { Readable } = require('stream');
+
+class MockReadable extends Readable {
+  constructor(options) {
+    super(options);
+  }
+  _read(size) { this.push(null); }
```
Comment (ewindisch) [inline __mocks__/v8-profiler-lambda.js]: comment that implementations of Readable push null to end the stream per nodejs docs

File: __mocks__/v8-profiler-lambda.js [hunk 9]
```diff
@@ -19,16 +29,29 @@ function stopProfiling() {
   return {
     export: cb => {
```
Comment (pselle) [inline __mocks__/v8-profiler-lambda.js]: Now that you're using `export()` vs `export(callback)`, you don't need the cb argument. I think this should've been caught by eslint, so maybe (separate issue) we should make sure that's set up

File: src/index.test.js [hunk 10]
```diff
@@ -53,5 +53,6 @@ test('works with iopipe', async function runTest() {
   expect(val).toBe('wow');
   expect(profilerRuntime.running).toBe(false);
   expect(putData.length).toBe(1);
-  expect(putData[0]).toBe('woot');
+  // Test that the data returned has the zip format magic bytes.
+  expect(putData[0].slice(0, 4)).toEqual(Buffer([80, 75, 3, 4]));
```
Comment (ewindisch) [inline src/index.test.js]: we could look inside the zip file and check the contents too

File: src/request.js [hunk 11]
```diff
@@ -15,7 +15,14 @@ export default function request(body, method, opts, authorizationHeader) {
     };
   }
   requestOptions['headers'] = requestOptions.headers || {};
-  requestOptions.headers['content-length'] = Buffer.byteLength(body);
+
+  /* Not a stream... */
+  if (!body.pipe) {
```
Comment (ewindisch) [inline src/request.js]: Remove this since we can't use streams here!

File: src/index.test.js (hunk 10 above)
Comment (ewindisch) [reply src/index.test.js]: We should replicate this test with the profiler enabled, the heap enabled, and with both enabled.

Maybe replicate this test with everything disabled as well! Then we know a disabled plugin doesn't break invocations.

File: src/index.js (hunk 4 above)
Comment (ewindisch) [reply src/index.js]: done.

File: src/index.js (hunk 6 above)
Comment (ewindisch) [reply src/index.js]: done

**Comment by coreylight** [comment]
Looking better @ewindisch! I'd like to see tests of the core functionality, probably utilizing the mock functionality that you started on with `takeSnapshot`

**Comment by ewindisch** [comment]
@coreylight more tests would be fantastic but seem to be outside the scope of this change. This adds tests to new functionality that are on par with existing tests. Expanded tests would be fantastic, but maybe as a future improvement?

**Comment by ewindisch** [comment]
This issue was blocked on changes to the signing API which have been completed. It has been updated to support file extensions. I'd love review! @kolanos

**Comment by coreylight** [comment]
Thanks @ewindisch, this is going to be awesome.

**Pull request merged**